### PR TITLE
Install layout builder and begin work on eCMS Layout

### PR DESCRIPTION
**Pull request by jordan-caldwell**
## Summary
This PR installs Layout Builder to the base profile and ports over a custom layout module that will be used to build out the eCMS Layout Builder functionality. The module has been converted to support only the feature eCMS needs and will be continued to be developed as new functionality arises. Theming will be included in a separate PR. 

A standard Drupal .editorconfig file has also been added.

## Metadata
| Question | Answer |
|----------|--------|
| Did you use a meaningful pull request title? | yes
| Did you apply meaningful labels to the pull request? | yes
| Did you perform a self review first? | yes
| Did you perform browser testing? | no
| Risk level | Low
| Relevant links | https://thinkoomph.jira.com/browse/RIG-14

### CODE REVIEW

File: ecms_base/modules/custom/ecms_layout/ecms_layout.module [hunk 1]
```diff
@@ -0,0 +1,90 @@
+<?php
+
+/**
+ * @file
+ * Lifespan Layout.
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/ecms_layout.module:5]: Remove the lifespan reference.

File: ecms_base/modules/custom/ecms_layout/ecms_layout.module [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ */
+function ecms_layout_preprocess_block__ecms_layout(array &$variables): void {
+  // Wrap the output of the block in a container.
+  $variables['content']['#type'] = 'container';
+}
+
+/**
+ * Implements hook_entity_operation().
+ */
+function lifespan_layout_entity_operation(EntityInterface $entity) {
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/ecms_layout.module:38]: Function name references lifespan.

Comment (pfrilling) [reply ecms_base/modules/custom/ecms_layout/ecms_layout.module:38]: Also, the return type should be `: array`.

File: ecms_base/modules/custom/ecms_layout/ecms_layout.module [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  $route_parameters = [
+    $entity_type_id => $entity->id(),
+  ];
+
+  // If current user has access to route, then add the operation link. The
+  // access check will only return TRUE if the bundle is Layout Builder-
+  // enabled, overrides are allowed, and user has necessary permissions.
+  $access_manager = \Drupal::service('access_manager');
+  if (!$access_manager->checkNamedRoute($route_name, $route_parameters, $account)) {
+    return;
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/ecms_layout.module:52]: Change this to `return [];` so they typehint is accurate.

File: ecms_base/modules/custom/ecms_layout/ecms_layout.module [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    ],
+  ];
+}
+
+/**
+ * Implements hook_plugin_filter_TYPE__CONSUMER_alter().
+ *
+ * Curate the blocks available in the Layout Builder "Add Block" UI.
+ */
+function ecms_layout_plugin_filter_block__layout_builder_alter(array &$definitions, array $extra) {
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/ecms_layout.module:69]: Return type hint here. `: void`.

File: ecms_base/modules/custom/ecms_layout/ecms_layout.module [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  $exclude_list = _ecms_layout_block_exclude_list();
+
+  // Remove blocks from users who do not have advanced permission.
+  if (\Drupal::currentUser()->hasPermission('administer ecms layout') == FALSE) {
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/ecms_layout.module:73]: Strict equality `===`.

File: ecms_base/modules/custom/ecms_layout/ecms_layout.permissions.yml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+administer ecms layout:
+  title: 'Administer eCMS layout'
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/ecms_layout.permissions.yml:2]: Newline at the end of the file.

File: ecms_base/modules/custom/ecms_layout/src/ECMSLayout.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\ECMS_layout;
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/src/ECMSLayout.php:5]: The namespace should be `Drupal\ecms_layout` to match the other classes below. I'd also rename the file to `EcmsLayout.php` and the class to `EcmsLayout`.

File: ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/FourColumnLayout.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\ecms_layout\Plugin\Layout;
+
+/**
+ * Provides a plugin class for one column layouts.
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/FourColumnLayout.php:8]: The comment references 'one column'. I think this should be 'four column'.

File: ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/LayoutBase.php [hunk 9]
```diff
@@ -0,0 +1,209 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\ecms_layout\Plugin\Layout;
+
+use Drupal\ecms_layout\ECMSLayout;
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/LayoutBase.php:7]: Change the class to `EcmsLayout` to match the above change.

File: ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/LayoutBase.php [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+    return $build;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration(): array {
+    return [
+      'background_color' => ECMSLayout::ROW_BACKGROUND_COLOR_NONE,
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/LayoutBase.php:53]: Class name to `EcmsLayout`.

File: ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/LayoutBase.php [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+  /**
+   * Get the background color options.
+   *
+   * @return array
+   *   The background color options.
+   */
+  protected function getBackgroundColorOptions(): array {
+    return [
+      ECMSLayout::ROW_BACKGROUND_COLOR_NONE => $this->t('None'),
+      ECMSLayout::ROW_BACKGROUND_COLOR_RED => $this->t('Red'),
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/LayoutBase.php:159]: Change the class name to `EcmsLayout`.

File: ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/ThreeColumnLayout.php [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Provides a plugin class for one column layouts.
+ */
+final class ThreeColumnLayout extends LayoutBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getColumnWidths(): array {
+    return [];
```
Comment (pfrilling) [inline ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/ThreeColumnLayout.php:16]: Should there be column options here?

**Review by pfrilling — CHANGES_REQUESTED** [review]
Found a few minor fixes.

File: ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/ThreeColumnLayout.php (hunk 12 above)
Comment (jordan-caldwell) [reply ecms_base/modules/custom/ecms_layout/src/Plugin/Layout/ThreeColumnLayout.php:16]: No, with three column options we do no not want to give them options for col size. They will all be 33% - 33% - 33%. This applies to 4 col layouts as well.

**Review by pfrilling — APPROVED** [review]

**Pull request merged**